### PR TITLE
feat: implement event lists for Page

### PR DESCRIPTION
**Pull request by aslushnikov**

### CODE REVIEW

File: packages/playwright-core/src/client/eventList.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  clear() {
+    this._list = [];
+  }
+
+  async take(optionsOrPredicate: WaitForEventOptions = {}): Promise<T> {
+    return await this._pageOrContext._wrapApiCall(async () => {
+      const predicate = typeof optionsOrPredicate === 'function' ? optionsOrPredicate : optionsOrPredicate.predicate;
+      while (true) {
+        const index = this._list.findIndex((e: T) => predicate ? !!predicate(e) : true);
```
Comment (dgozman) [inline packages/playwright-core/src/client/eventList.ts:51]: `await predicate(e)`

File: packages/playwright-core/src/client/eventList.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  clear() {
+    this._list = [];
+  }
+
+  async take(optionsOrPredicate: WaitForEventOptions = {}): Promise<T> {
+    return await this._pageOrContext._wrapApiCall(async () => {
+      const predicate = typeof optionsOrPredicate === 'function' ? optionsOrPredicate : optionsOrPredicate.predicate;
+      while (true) {
```
Comment (dgozman) [inline packages/playwright-core/src/client/eventList.ts:50]: Needs a dead line.

File: packages/playwright-core/src/client/eventList.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        const index = this._list.findIndex((e: T) => predicate ? !!predicate(e) : true);
+        if (index !== -1) {
+          const [element] = this._list.splice(index, 1);
+          return element;
+        }
+        await this._pageOrContext.waitForEvent(this._eventName, optionsOrPredicate);
```
Comment (dgozman) [inline packages/playwright-core/src/client/eventList.ts:56]: Let's add a test for the while(true).

File: tests/page/page-eventlists.spec.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { test as it, expect } from './pageTest';
+
+it('should not collect events by default', async ({ page, server }) => {
```
Comment (dgozman) [inline tests/page/page-eventlists.spec.ts:18]: I'd remove first two tests as useless.

File: tests/page/page-eventlists.spec.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    page.evaluate(() => {
+      console.log('foo');
+      console.log('bar');
+    }),
+  ]);
+  expect(msg1.text()).toBe('foo');
+  expect(msg2.text()).toBe('bar');
+});
+
+it('eventList.take should respect timeout option', async ({ page, server }) => {
```
Comment (dgozman) [inline tests/page/page-eventlists.spec.ts:84]: Let's do a test for `take(predicate)` and make it resolve only after the second console message.

File: packages/playwright-core/src/client/eventList.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    return await this._pageOrContext._wrapApiCall(async () => {
+      const predicate = typeof optionsOrPredicate === 'function' ? optionsOrPredicate : optionsOrPredicate.predicate;
+      const timeout = this._pageOrContext._timeoutSettings.timeout(typeof optionsOrPredicate === 'function' ? {} : optionsOrPredicate);
+      const waiter = new Waiter(this._pageOrContext, this._eventName, apiName);
+      this._pageOrContext._setupWaiter(waiter, this._eventName, timeout, `waiting for ${apiName}`);
+
+      const composedPredicate = predicate ? (arg: WrappedEvent<T>) => predicate(arg.event) : undefined;
+
+      while (true) {
+        const uniqueObj = await waiter.waitForEvent(this._pageOrContext, this._eventName, composedPredicate, this._list);
```
Comment (dgozman) [inline packages/playwright-core/src/client/eventList.ts:62]: Let's comment that this line will throw after timeout/close.

File: packages/playwright-core/src/client/eventList.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        const index = this._list.indexOf(uniqueObj);
```
Comment (dgozman) [inline packages/playwright-core/src/client/eventList.ts:63]: `waiter` might not return the wrapped event, make sure there's a test for this.

File: docs/src/api/class-pageevents.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
+# class: PageEvents
```
[the remark this answers is not in the document]
Comment (aslushnikov) [reply docs/src/api/class-pageevents.md:1]: let's bikeshed as this lands

File: docs/src/api/class-eventlist.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+## method: EventList.track
+* since: v1.28
+
+Enables event tracking.
+
+## method: EventList.untrack
+* since: v1.28
+
+Disables event tracking.
```
[the remark this answers is not in the document]
Comment (aslushnikov) [reply docs/src/api/class-eventlist.md:70]: let's discuss on API review

File: docs/src/api/class-eventlist.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+```js
+// Before:
+const [consoleMessage] = await Promise.all([
+  page.waitForEvent('console'),
+  page.getByText('Log Console Message').click(),
+]);
+
+// After:
+page.events.console.track();
```
[the remark this answers is not in the document]
Comment (aslushnikov) [reply docs/src/api/class-eventlist.md:23]: > Make separate js-ts vs js-library snippets, and omit track() in the testing snippet.

This would make sense once we engage in playwright test by default

File: docs/src/api/class-eventlist.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+## async method: EventList.take
+* since: v1.28
+- returns: <[any]>
+
+Returns the first event that satisfies condition, if any. If no condition
+is given, returns the first accumulated event.
+
+```js
+page.events.console.track();
```
[the remark this answers is not in the document]
Comment (aslushnikov) [reply docs/src/api/class-eventlist.md:47]: i'll do in a follow-up

File: docs/src/api/class-eventlist.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+* since: v1.28
+
+Event list collects events of a certain kind for future use; instances
+of this class are accessible via the [`property: Page.events`] getter.
+
+Notice that event lists simply store all historic events of a given type and
+might store outdated objects, e.g. `page.events.popup.all()` might return
+already-closed pages.
+
+Use event lists instead of a `Promise.all` pattern. Consider the
```
Comment (dgozman) [inline docs/src/api/class-eventlist.md:12]: I'd remove any references of `Promise.all`.

File: docs/src/api/class-eventlist.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+following example:
+
+```js
+page.events.filechooser.track();
+await page.getByText('Pick a file').click();
+const fileChooser = await page.events.filechooser.take();
```
Comment (dgozman) [inline docs/src/api/class-eventlist.md:18]: - `await fileChooser.setInputFiles(...)`
- perhaps not capitalize it?

File: docs/src/api/class-eventlist.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+Returns the first event that satisfies condition, if any. If no condition
+is given, returns the first accumulated event. If nothing matches, then
+the method will wait for a matching event to come.
+
+```js
+page.events.console.track();
+page.events.console.clear();
+await page.getByText('Log Console Message').click();
+const consoleMessage = await page.events.console.take();
```
Comment (dgozman) [inline docs/src/api/class-eventlist.md:44]: I'd rather reuse filechooser example.

File: docs/src/test-api/class-testconfig.md [hunk 15]
```diff
@@ -656,7 +656,7 @@ export default config;
   - `port` ?<[int]> The port that your http server is expected to appear on. It does wait until it accepts connections. Exactly one of `port` or `url` is required.
   - `url` ?<[string]> The url on your http server that is expected to return a 2xx, 3xx, 400, 401, 402, or 403 status code when the server is ready to accept connections. Exactly one of `port` or `url` is required.
   - `ignoreHTTPSErrors` ?<[boolean]> Whether to ignore HTTPS errors when fetching the `url`. Defaults to `false`.
-  - `timeout` ?<[int]> How long to wait for the process to start up and be available in milliseconds. The same timeout is also used to terminate the process. Defaults to 60000.
+  - `timeout` ?<[int]> How long to wait for the process to start up and be available in milliseconds. Defaults to 60000.
```
Comment (dgozman) [inline docs/src/test-api/class-testconfig.md:659]: Seems unrelated.

**Review by dgozman — APPROVED** [review]

**Comment by pavelfeldman** [comment]
GC
